models/dbfs/dbfs.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+package dbfs
+
+import (
+	"context"
+	"os"
+
+	"code.gitea.io/gitea/models/db"
+)
+
+type dbfsMeta struct {
+	ID              int64  `xorm:"pk autoincr"`
+	FullPath        string `xorm:"VARCHAR(500) UNIQUE NOT NULL"`
+	BlockSize       int64  `xorm:"BIGINT NOT NULL"`
+	FileSize        int64  `xorm:"BIGINT NOT NULL"`
+	CreateTimestamp int64  `xorm:"BIGINT NOT NULL"`
+	ModifyTimestamp int64  `xorm:"BIGINT NOT NULL"`
+}
+
+type dbfsData struct {
+	ID         int64  `xorm:"pk autoincr"`
+	Revision   int64  `xorm:"BIGINT NOT NULL"`
+	MetaID     int64  `xorm:"BIGINT index(meta_offset) NOT NULL"`
+	BlobOffset int64  `xorm:"BIGINT index(meta_offset) NOT NULL"`
+	BlobSize   int64  `xorm:"BIGINT NOT NULL"`
+	BlobData   []byte `xorm:"BLOB NOT NULL"`
+}
```

Comment:
Feel free to unresolve if you have any more ideas.

Reply:
@wxiaoguang regarding https://github.com/go-gitea/gitea/pull/21937#issuecomment-1409029620:
Ah, I thought I posted it here as well, but apparently, this was only in a private chat.
Here is said alternative based on the `logIndex+logContent` approach mentioned above.
<details>
<summary>DBFS alternative</summary>

```go
(global) var taskLogLock sync.Mutex // TODO for later: use a `sync.Mutex` per taskID instead to allow parallel upload of different logs

…
func InsertTaskLogs(taskID int64, logs []*TaskLog) (int64, error) {

  // Block other threads from uploading in parallel to ensure a consistent outcome
  taskLogLock.Lock()
  defer taskLogLock.Unlock()

  // Calculate next free log line number
  var previous *TaskLine
  has, err := e.Where("task_id=?", taskID).Desc("line_number").Get(&previous)
  nextLineIndex := 0 // Alternatively 1 if we want to count one-based, which would also be valid
  
  if err != nil {
    return err
  } else if has {
    nextLineIndex = previous.LineNumber + 1
  }

  // Insert the next n log-lines
  for i, log := range logs {
    e.Insert(&TaskLine{LineNumber: i + nextLineIndex, …})
  }
```

The response was along the lines of "this will most likely work too, but the PR should be merged soon so we'll postpone it and post the PR to get rid of it afterward"

</details>

Reply:
It's carefully designed, I think wolfogre has explained the reasons. 

And I can add more reasons about why `logIndex+logLine` approach is not used and why DBFS wins:

* It's easier to upload the logs to other storages. In the end, Gitea needs to read logs from external storages like S3. `logIndex+logLine`  would make the reading code totally different. While DBFS and S3 can have the same stream interface.
* If you store too many rows in a table, it's not friendly to databases. Think about you are running 1000 tasks and each task have 1000 lines, you will get 1000,000 rows in your table.